salt/_utils/kubernetes_utils.py

```diff
@@ -0,0 +1,835 @@
+
+
+_CUSTOM_KINDS = [
+    CRKindInfo('monitoring.coreos.com/v1', 'Alertmanager',
+               scope='namespaced', plural='alertmanagers'),
+    CRKindInfo('monitoring.coreos.com/v1', 'Prometheus',
+               scope='namespaced', plural='prometheuses'),
+    CRKindInfo('monitoring.coreos.com/v1', 'PrometheusRule',
+               scope='namespaced', plural='prometheusrules'),
+    CRKindInfo('monitoring.coreos.com/v1', 'ServiceMonitor',
+               scope='namespaced', plural='servicemonitors'),
+]
+
+KNOWN_CUSTOM_KINDS = {kind.key: kind for kind in _CUSTOM_KINDS}
+
+
+class _DictWrapper(object):
+    """Wrapper for dynamic attribute access support over a simple dict.
+
+    Implemented by recursively casting all dicts within the original dict as
+    other `_DictWrapper`s, and overloading the `__getattribute__` to fallback
+    on the internal `_fields` for retrieving an attribute.
+
+    Used by `CustomObject` to emulate the other models for `kubernetes.client`.
+
+    >>> example = _DictWrapper({
+    ...     'a': {'b': 'value'},
+    ...     'c': [{'subkey': 'subvalue'}]
+    ... })
+    >>> example.a.b
+    'value'
+    >>> example.c[0].subkey
+    'subvalue'
+    >>> example.unknown
+    Traceback (most recent call last):
+        ...
+    AttributeError: Custom object has no attribute 'unknown'
+    """
+    def __init__(self, fields):
+        self._fields = fields
+
+    @classmethod
+    def from_value(cls, value):
+        if isinstance(value, dict):
+            return cls(value)
+        if isinstance(value, list):
+            return [cls.from_value(val) for val in value]
+        return value
+
+    def __repr__(self):
+        return repr(self._fields)
+
+    def __getattribute__(self, name):
+        try:
+            return super(_DictWrapper, self).__getattribute__(name)
+        except AttributeError:
+            if name in self._fields:
+                return self.from_value(self._fields[name])
+            for key in self._fields:
+                if _convert_attribute_name(key) == name:
+                    return self.from_value(self._fields[key])
+            raise AttributeError(
+                "Custom object has no attribute '{}'".format(name)
+            )
+
+    def __setattr(self, name, value):
+        # First check for class values then retrieve from dict
+        if name in ['_fields']:
+            super(_DictWrapper, self).__setattr__(name, value)
+        else:
+            self._fields[name] = value
+
+
+class CustomObject(object):
+    """Helper class to generate a "model" interface for CustomResources.
+
+    The source manifest is converted to use snake case for all its keys and
+    sub-keys, as is done in the `kubernetes.client`.
+
+    This fake "model" also uses a `_DictWrapper` internally to provide dynamic
+    attribute access on the instance's `body`.
+    """
+    def __init__(self, manifest):
+        self._body = manifest
+        self._attr_dict = _DictWrapper(self._body)
```

Comment:
That should be `_DictWrapper(_cast_dict_keys(manifest, _convert_attribute_name))`.

For a reminder of the rationale of all this, we're trying to do the same between standard objects in `kubernetes.client` and those `CustomObject`s.

Reply:
> For a reminder of the rationale of all this, we're trying to do the same between standard objects in `kubernetes.client` and those `CustomObject`s.

Yes I know but we can't as for custom object we need to call with a dict containing all value in camel case so several choice:
- we convert the dict into an object that look like "kubernetes.client" objects we do some tuning directly in that object and then we need to cast it back to camel case dict (ugly IMO and we need something to convert from snake case to camel case :confused: )
- we keep a dict in camel case every where and so we need to do a distinction between custom object and "python-kubernetes objects" in the module (doable and not a big effort)
- we keep the initial dict and we have an "helper" (DictWrapper) that when you retrieve value first try with camel case and then snake case so that we can manipulate both object in the same way in the module (current choice)

Reply:
Basically it's a dict so execpt if we edit directly the full dict `self._body == self._attr_dict._fields` but yes agree it's cleaner to use `self._attr_dict._fields`